crates/runtime/src/instance/allocator/pooling.rs

```diff
@@ -0,0 +1,1714 @@
+        // See doc comments for `ModuleLimits` for these default values
+        Self {
+            imported_functions: 1000,
+            imported_tables: 0,
+            imported_memories: 0,
+            imported_globals: 0,
+            types: 100,
+            functions: 10000,
+            tables: 1,
+            memories: 1,
+            globals: 10,
+            table_elements: 10000,
+            memory_pages: 160,
+        }
+    }
+}
+
+/// Represents the limits placed on instances by the pooling instance allocator.
+#[derive(Debug, Copy, Clone)]
+pub struct InstanceLimits {
+    /// The maximum number of concurrent instances supported (default is 1000).
+    pub count: u32,
+
+    /// The maximum size, in bytes, of host address space to reserve for each linear memory of an instance.
+    ///
+    /// Note: this value has important performance ramifications.
+    ///
+    /// On 64-bit platforms, the default for this value will be 6 GiB.  A value of less than 4 GiB will
+    /// force runtime bounds checking for memory accesses and thus will negatively impact performance.
+    /// Any value above 4 GiB will start eliding bounds checks provided the `offset` of the memory access is
+    /// less than (`memory_reservation_size` - 4 GiB).  A value of 8 GiB will completely elide *all* bounds
+    /// checks; consequently, 8 GiB will be the maximum supported value. The default of 6 GiB reserves
+    /// less host address space for each instance, but a memory access with an offet above 2 GiB will incur
+    /// runtime bounds checks.
+    ///
+    /// On 32-bit platforms, the default for this value will be 10 MiB. A 32-bit host has very limited address
+    /// space to reserve for a lot of concurrent instances.  As a result, runtime bounds checking will be used
+    /// for all memory accesses.  For better runtime performance, a 64-bit host is recommended.
+    ///
+    /// This value will be rounded up by the WebAssembly page size (64 KiB).
+    pub memory_reservation_size: u64,
+}
+
+impl Default for InstanceLimits {
+    fn default() -> Self {
+        // See doc comments for `InstanceLimits` for these default values
+        Self {
+            count: 1000,
+            #[cfg(target_pointer_width = "32")]
+            memory_reservation_size: 0xA00000,
+            #[cfg(target_pointer_width = "64")]
+            memory_reservation_size: 0x180000000,
+        }
+    }
+}
+
+/// The allocation strategy to use for the pooling instance allocator.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum PoolingAllocationStrategy {
+    /// Allocate from the next available instance.
+    NextAvailable,
+    /// Allocate from a random available instance.
+    Random,
+}
+
+impl PoolingAllocationStrategy {
+    fn next(&self, free_count: usize) -> usize {
+        debug_assert!(free_count > 0);
+
+        match self {
+            Self::NextAvailable => free_count - 1,
+            Self::Random => rand::thread_rng().gen_range(0, free_count),
```

Comment:
This is intended to be used in production environments to make it harder for attackers to predict linear memory addresses (linear memory ASLR, if you will).

In Lucet, there's `Linear` (which I'm calling `NextAvailable` here), `Random` (same implementation as here), and `CustomRandom(Arc<Mutex<dyn RngCore>>)` which is used in testing to simulate randomness (not implemented in this PR).

Perhaps we could include `CustomRandom` just in case a different rng is desired?  Thread-local variants would definitely be preferred, though, as the only locks taken in the pooling allocator currently are for modifying the free lists.

Reply:
Ah ok makes sense, let's stick with just using `thread_rng` for now (no need I think yet to start using a custom rng unless necessary), and if that's a problem in the future we can fix it then.